### PR TITLE
Isolated processing improvements

### PR DESCRIPTION
**Pull request by ViacheslavKlimov**
## Pull Request description

image

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: ui-ngx/src/app/modules/home/components/profile/tenant-profile.component.ts [hunk 1]
```diff
@@ -76,6 +76,52 @@ export class TenantProfileComponent extends EntityComponent<TenantProfile> {
         additionalInfo: {
           description: ''
         }
+      },
+      {
+        id: guid(),
+        name: 'HighPriority',
+        topic: 'tb_rule_engine.hp',
+        pollInterval: 25,
```
Comment (smatvienko-tb) [inline ui-ngx/src/app/modules/home/components/profile/tenant-profile.component.ts:84]: I would like to set pool interval as lazy as possible.

Even 1000ms for 500 tenants*2partitions*(main,hp,sq)3queues = 3000 poll per second for the cluster. Please estimate the load before provide such a parameter value

File: ui-ngx/src/app/modules/home/components/profile/tenant-profile.component.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ export class TenantProfileComponent extends EntityComponent<TenantProfile> {
         consumerPerPartition: true,
         name: 'Main',
         packProcessingTimeout: 2000,
-        partitions: 10,
+        partitions: 2,
```
Comment (smatvienko-tb) [inline ui-ngx/src/app/modules/home/components/profile/tenant-profile.component.ts:62]: I would like to control a **retention policy** for isolated topic.
With 3k topics with a default limit 1Gb and 7 days, we will end up with **3 TeraByte** * replication factor PVC size !!!!
Retention policy like this
image

That is a lot!

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java [hunk 3]
```diff
@@ -181,9 +186,20 @@ public void removeQueue(TransportProtos.QueueDeleteMsg queueDeleteMsg) {
     @Override
     public TopicPartitionInfo resolve(ServiceType serviceType, String queueName, TenantId tenantId, EntityId entityId) {
         TenantId isolatedOrSystemTenantId = getIsolatedOrSystemTenantId(serviceType, tenantId);
+        if (queueName == null) {
+            queueName = MAIN_QUEUE_NAME;
+        }
         QueueKey queueKey = new QueueKey(serviceType, queueName, isolatedOrSystemTenantId);
         if (!partitionSizesMap.containsKey(queueKey)) {
-            queueKey = new QueueKey(serviceType, isolatedOrSystemTenantId);
+            if (isolatedOrSystemTenantId.isSysTenantId()) {
```
Comment (smatvienko-tb) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java:194]: No test - no merge. This is a key feature :)

**Review by smatvienko-tb — CHANGES_REQUESTED** [review]
Good job! Please, take a look at my comments.

File: application/src/main/java/org/thingsboard/server/service/entitiy/queue/DefaultTbQueueService.java [hunk 4]
```diff
@@ -108,21 +106,15 @@ private void onQueueUpdated(Queue queue, Queue oldQueue) {
...
-
-                scheduler.schedule(() -> {
-                    for (int i = currentPartitions; i < oldPartitions; i++) {
-                        String fullTopicName = new TopicPartitionInfo(queue.getTopic(), queue.getTenantId(), i, false).getFullTopicName();
-                        log.info("Removed partition [{}]", fullTopicName);
-                        tbQueueAdmin.deleteTopic(
-                                fullTopicName);
-                    }
-                }, DELETE_DELAY, TimeUnit.SECONDS);
+                // TODO: move all the messages left in old partitions and delete topics
```
[the remark this answers is not in the document]
Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/entitiy/queue/DefaultTbQueueService.java:117]: Please add more logging to the ```updateQueuesByTenants```. I am not sure it works very well.

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java [hunk 5]
```diff
@@ -181,9 +186,20 @@ public void removeQueue(TransportProtos.QueueDeleteMsg queueDeleteMsg) {
...
+                queueKey = new QueueKey(serviceType, TenantId.SYS_TENANT_ID);
+            } else {
+                queueKey = new QueueKey(serviceType, queueName, TenantId.SYS_TENANT_ID);
```
Comment (ashvayka) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java:197]: Please add TODO to improve TransportService.getQueueRoutingInfo for cases when it does not fit into the reply (kafka limitations). No need to fix this now. But a TODO is always great :-)

File: application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java [hunk 6]
```diff
@@ -107,13 +114,14 @@ public class DefaultTbRuleEngineConsumerService extends AbstractConsumerService<
...
-    //    private final TenantId tenantId;
+    private final TbQueueProducerProvider producerProvider;
+    private final TbQueueAdmin queueAdmin;
     private final ConcurrentMap<QueueKey, TbQueueConsumer<TbProtoQueueMsg<ToRuleEngineMsg>>> consumers = new ConcurrentHashMap<>();
     private final ConcurrentMap<QueueKey, Queue> consumerConfigurations = new ConcurrentHashMap<>();
     private final ConcurrentMap<QueueKey, TbRuleEngineConsumerStats> consumerStats = new ConcurrentHashMap<>();
     private final ConcurrentMap<QueueKey, TbTopicWithConsumerPerPartition> topicsConsumerPerPartition = new ConcurrentHashMap<>();
     final ExecutorService submitExecutor = Executors.newSingleThreadExecutor(ThingsBoardThreadFactory.forName("tb-rule-engine-consumer-submit"));
-    final ScheduledExecutorService repartitionExecutor = Executors.newScheduledThreadPool(1, ThingsBoardThreadFactory.forName("tb-rule-engine-consumer-repartition"));
+    final ScheduledExecutorService repartitionExecutor = Executors.newScheduledThreadPool(2, ThingsBoardThreadFactory.forName("tb-rule-engine-consumer-repartition"));
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java:124]: Why? The idea is to have one thread to make sure all repartition evens are sync.

File: application/src/main/java/org/thingsboard/server/actors/app/AppActor.java [hunk 7]
```diff
@@ -123,7 +124,11 @@ private void initTenantActors() {
         try {
             if (systemContext.isTenantComponentsInitEnabled()) {
                 PageDataIterable<Tenant> tenantIterator = new PageDataIterable<>(tenantService::findTenants, ENTITY_PACK_LIMIT);
+                Set<UUID> assignedProfiles = systemContext.getServiceInfoProvider().getAssignedTenantProfiles();
                 for (Tenant tenant : tenantIterator) {
+                    if (!assignedProfiles.isEmpty() && !assignedProfiles.contains(tenant.getTenantProfileId().getId())) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/actors/app/AppActor.java:129]: Please take a look at ```onToDeviceActorMsg```. We don't need to dispatch those messages for tenants processed on different servers. Otherwise, the TenantActor will be created. See all usages of  getOrCreateTenantActor

Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/actors/app/AppActor.java:129]: Much easier to read is 

```java
if (assignedProfiles.isEmpty() || assignedProfiles.contains(tenant.getTenantProfileId().getId())) {
    log.debug("[{}] Creating tenant actor", tenant.getId());
    getOrCreateTenantActor(tenant.getId());
    log.debug("[{}] Tenant actor created.", tenant.getId());
}

```

File: common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java [hunk 8]
```diff
@@ -423,18 +440,51 @@ protected ServiceInfo resolveByPartitionIdx(List<ServiceInfo> servers, QueueKey
...
+                    profileId = routingInfo.getProfileId();
+                } else {
+                    profileId = null;
+                }
+
+                List<ServiceInfo> responsible = responsibleServices.get(profileId);
+                if (responsible == null) {
+                    // if there are no dedicated servers for this tenant profile, or for system queues,
+                    // using the servers that are not responsible for any profile
+                    responsible = servers.stream()
```
Comment (ashvayka) [inline common/queue/src/main/java/org/thingsboard/server/queue/discovery/HashPartitionService.java:458]: I don't like the idea of manipulating the ```responsibleServices``` outside the ```recalculateParititions```. You may just create new variable and update it in the ```recalculateParititions```. Somethings like ```commonServices``` - list of services that process data from generic tenant profiles.

File: application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java [hunk 9]
```diff
@@ -156,7 +156,7 @@ public void init() {
         super.init("tb-rule-engine-consumer", "tb-rule-engine-notifications-consumer");
         List<Queue> queues = queueService.findAllQueues();
         for (Queue configuration : queues) {
-            initConsumer(configuration);
+            initConsumer(configuration); // TODO: if this Rule Engine is assigned specific profile, don't init other consumers and properly handle queue update events
```
[the remark this answers is not in the document]
Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java:159]: If we init all the consumers on all the services, then we may have a complete mess, right? Or do I miss something?

File: application/src/main/java/org/thingsboard/server/actors/app/AppActor.java (hunk 7 above)
Comment (smatvienko-tb) [reply application/src/main/java/org/thingsboard/server/actors/app/AppActor.java:129]: Agree. The conditions with NOT is hard to follow

File: application/src/main/java/org/thingsboard/server/actors/app/AppActor.java [hunk 10]
```diff
@@ -181,31 +183,37 @@ private void onComponentLifecycleMsg(ComponentLifecycleMsg msg) {
...
             }
         }
     }
 
-    private TbActorRef getOrCreateTenantActor(TenantId tenantId) {
-        return ctx.getOrCreateChildActor(new TbEntityActorId(tenantId),
-                () -> DefaultActorService.TENANT_DISPATCHER_NAME,
-                () -> new TenantActor.ActorCreator(systemContext, tenantId));
+    private Optional<TbActorRef> getOrCreateTenantActor(TenantId tenantId) {
+        if (systemContext.getServiceInfoProvider().isService(ServiceType.TB_CORE) ||
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/actors/app/AppActor.java:201]: I suggest taking this if into the lambda and passing it to ctx.getOrCreateChildActor. Because if the actor is there, then there is no point in checking all these conditions, it is enough to just return it.

File: application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java [hunk 11]
```diff
@@ -156,7 +156,9 @@ public void init() {
         super.init("tb-rule-engine-consumer", "tb-rule-engine-notifications-consumer");
         List<Queue> queues = queueService.findAllQueues();
         for (Queue configuration : queues) {
-            initConsumer(configuration); // TODO: if this Rule Engine is assigned specific profile, don't init other consumers and properly handle queue update events
+            if (partitionService.isManagedByCurrentService(configuration.getTenantId())) {
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/queue/DefaultTbRuleEngineConsumerService.java:159]: :+1:
